src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadata.cs

```diff
@@ -427,6 +430,94 @@ public override bool ValidateChildren
             }
         }
 
+        /// <inheritdoc />
+        public override bool IsValidationRequired
+        {
+            get
+            {
+                if (!_isValidationRequired.HasValue)
+                {
+                    CalculateIsValidationRequired();
+                }
+
+                Debug.Assert(_isValidationRequired.HasValue);
+                return _isValidationRequired.Value;
+            }
+        }
+
+        private void CalculateIsValidationRequired()
+        {
+            var visited = new HashSet<ModelMetadata>();
+
+            if (TryVisit(this, out var requiresValidation))
+            {
+                _isValidationRequired = requiresValidation;
+            }
+            else
+            {
+                // Assume true by default.
+                _isValidationRequired = true;
+            }
+
+            bool TryVisit(ModelMetadata metadata, out bool result)
+            {
+                RuntimeHelpers.EnsureSufficientExecutionStack();
+
+                if (!visited.Add(metadata))
+                {
+                    // A cycle. Return the value as is without attempting to calculate it now. We'll calculate it at a later point.
+                    result = default;
+                    return false;
+                }
+
+                if (!(metadata is DefaultModelMetadata defaultModelMetadata))
+                {
+                    // We cannot inspect this ModelMetadata. Assume it requires validation.
+                    result = true;
+                    return true;
+                }
+
+                if (defaultModelMetadata._isValidationRequired.HasValue)
+                {
+                    result = defaultModelMetadata._isValidationRequired.Value;
+                    return true;
+                }
+
+                if (defaultModelMetadata.ValidationMetadata.IsValidationRequired)
```

Comment:
I don't think you can do this. Imagine the following structure:
```
public class Person
{
    public Dog BestFriend { get; set; }
}

public class Dog
{
    public Person BestFriend { get; set; }
}
```

I expected to see code where you recursively call `TryVisit`